### PR TITLE
Allow rebase in merge tree outside resubmit

### PR DESCRIPTION
**Pull request by anthony-murphy**
# Description

The current reconnect logic of the merge tree assumes that all outstanding operations will be rebased for reconnect, and they will be done in order of submission. This is always true in the reconnection case as resbumit is only called after the loader and runtime is sure it has seen everything it will from the previous client id, as only then is it safe to resend ops, otherwise we can end up with duplicate ops.

However, if we want to rebase unsent ops that current assumptions break, specifically that all outstanding ops will be rebased. As now there can be inflight ops that were generated and sent before the rebase point. To account for this this change will modify the logic such that on first rebase, the rebase point is selected, specifically the segment group. We will take all outstanding segment groups after this rebase point and stash them. We will enforce that all changes after the rebase point will be rebased. This is requirement, as any change after the rebase point becomes invalid when the changes before it have been rebased. There should also be no new changes interleaved during the rebase process.

related [AB#4757](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4757) [AB#4768](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4768)

### CODE REVIEW

File: packages/dds/merge-tree/src/collections/list.ts [hunk 1]
```diff
@@ -135,7 +171,7 @@ export class List<T>
 		return node instanceof DataNode && node.headNode === this.headNode;
 	}
 
-	remove(node: ListNode<T> | undefined): ListNode<T> | undefined {
+	private _remove(node: ListNode<T> | undefined): DataNode<T> | undefined {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/collections/list.ts:174]: this just lets me re-use the type narrowing in in splice we get from _includes

**Comment by vladsud** [comment]
@anthony-murphy , can you please describe what is the problem and what is the fix?
Description seems to be a template :)

**Comment by anthony-murphy** [comment]
> If we want to move into this direction, then I think DDS stress test framework (by @Abe27342) should support testing this condition (partial rebase of pending ops). Probably an issue for another PR, but it's hard to see us moving forward without having that capability.

As I understood the original plan for batched groups we were going to implement the necessary behavior off by default, and then roll it out dds by dds and fix issues. So it seems like this kind of testing should already be a part of the plan, and if not should be added.

**Review by andre4i — APPROVED** [review]

**Comment by andre4i** [comment]
Merging this in order to get some early feedback from partners. The feature is disabled by default, see https://github.com/microsoft/FluidFramework/pull/16176

**Pull request merged**